Sources/X509/TinyArray.swift

```diff
@@ -0,0 +1,230 @@
+            // newElements just contains a single element
+            // and we hit the fast path
+            self = .one(firstElement)
+            return
+        }
+        
+        var elements: [Element] = []
+        elements.reserveCapacity(newElements.underestimatedCount)
+        elements.append(firstElement)
+        elements.append(secondElement)
+        while let nextElement = try iterator.next()?.get() {
+            elements.append(nextElement)
+        }
+        self = .arbitary(elements)
+    }
+    
+    @inlinable
+    init() {
+        self = .arbitary([])
+    }
+    
+    @inlinable
+    mutating func append(_ newElement: Element) {
+        self.append(contentsOf: CollectionOfOne(newElement))
+    }
+    
+    @inlinable
+    mutating func append(contentsOf newElements: some Sequence<Element>){
+        switch self {
+        case .one(let firstElement):
+            var iterator = newElements.makeIterator()
+            guard let secondElement = iterator.next() else {
+                // newElements is empty, nothing to do
+                return
+            }
+            var elements: [Element] = []
+            elements.reserveCapacity(1 + newElements.underestimatedCount)
+            elements.append(firstElement)
+            elements.append(secondElement)
+            elements.appendRemainingElements(from: &iterator)
+            self = .arbitary(elements)
+            
+        case .arbitary(var elements):
+            if elements.isEmpty {
+                if let array = newElements as? Array<Element> {
+                    if array.count == 1 {
+                        self = .one(array[0])
+                    } else {
+                        self = .arbitary(array)
+                    }
+                } else {
+                    // if `self` is currently empty and `newElements` just contains a single
+                    // element, we skip allocating an array and set `self` to `.one(firstElement)`
+                    var iterator = newElements.makeIterator()
+                    guard let firstElement = iterator.next() else {
+                        // newElements is empty, nothing to do
+                        return
+                    }
+                    guard let secondElement = iterator.next() else {
+                        // newElements just contains a single element
+                        // and we hit the fast path
+                        self = .one(firstElement)
+                        return
+                    }
+                    elements.reserveCapacity(elements.count + newElements.underestimatedCount)
+                    elements.append(firstElement)
+                    elements.append(secondElement)
+                    elements.appendRemainingElements(from: &iterator)
+                    self = .arbitary(elements)
+                }
+            } else {
+                elements.append(contentsOf: newElements)
```

Comment:
I've found the compiler to be very hit and miss with avoiding CoWs in these situations. Might be worth verifying it doesn't CoW or temporarily setting `self = .arbitrary([])`

Reply:
Could you review https://github.com/apple/swift-certificates/pull/96 as well? Then I can write an allocation test as well

Reply:
Added an allocation test in https://github.com/apple/swift-certificates/pull/98/commits/4f1037b5469a63ef89af5c8104690c35fd4c2394 that shows that we only trigger a single allocation and do not trigger CoW on `append(_:)` which calls down to `append(contentsOf:)`